// lint-md/lib/lint-rules/helper/string.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.endSpaceLen = exports.startSpaceLen = exports.substr = exports.subErrorStr = exports.stringType = exports.charType = void 0;

var isChinese = function isChinese(s) {
  return /^[\u4e00-\u9fa5]{1}$/.test(s);
};

var isAlphabet = function isAlphabet(s) {
  return /^[a-zA-Z]{1}$/.test(s);
};

var isNumber = function isNumber(s) {
  return /^[0-9]{1}$/.test(s);
};
/**
 * 将字符映射成标记字符
 * @param c
 * @returns {string}
 */


var charType = function charType(c) {
  return isNumber(c) ? 'N' : isAlphabet(c) ? 'A' : isChinese(c) ? 'Z' : '-';
};

exports.charType = charType;

var stringType = function stringType(s) {
  return s.split('').map(function (s) {
    return charType(s);
  }).join('');
};
/**
 * 显示从给定位置为原点，一定范围内的字符串片段
 * @param {String} s         给定字符串
 * @param {Number} index     原点位置
 * @param {Number} length    获取字符串长度
 */


exports.stringType = stringType;

var subErrorStr = function subErrorStr(s, index, length) {
  return s.substr(Math.max(index - Math.floor(length / 2), 0), length);
};

exports.subErrorStr = subErrorStr;

var substr = function substr(s) {
  var len = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 12;
  if (!s) return s;
  return s.length > len ? "".concat(s.substr(0, len), "...") : s;
};
/**
 * 获取字符串开头的空格数量
 * @param {String} str
 */


exports.substr = substr;

var startSpaceLen = function startSpaceLen(str) {
  var len = 0;
  var array = str.split('');

  for (var index = 0; index < array.length; index++) {
    if (array[index] === ' ') {
      len++;
    } else {
      break;
    }
  }

  return len;
};
/**
 * 获取字符串结尾的空格数量
 * @param {String} str
 */


exports.startSpaceLen = startSpaceLen;

var endSpaceLen = function endSpaceLen(str) {
  var len = 0;
  var array = str.split('');

  for (var index = array.length - 1; index >= 0; index--) {
    if (array[index] === ' ') {
      len++;
    } else {
      break;
    }
  }

  return len;
};

exports.endSpaceLen = endSpaceLen;
