// lint-md/lib/lint-rules/use-standard-ellipsis.js
"use strict";

var _astPlugin = require("ast-plugin");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var _require = require('./helper/string'),
    subErrorStr = _require.subErrorStr;

var showLength = 14; // 找到所有的 …

var findAllSingleEllipsis = function findAllSingleEllipsis(s) {
  var r = [];
  var re = /…{1,}/g; // 使用正则匹配

  while (true) {
    var matched = re.exec(s); // 只要不是两个，都是不规范的

    if (matched && matched[0].length !== 2) {
      r.push({
        index: matched.index,
        length: matched[0].length
      });
    } else {
      break;
    }
  }

  return r;
}; // 找到所有的 . 组成的省略号


var findAllDotEllipsis = function findAllDotEllipsis(s) {
  var r = [];
  var re = /\.{4,}/g; // 使用正则匹配

  while (true) {
    var matched = re.exec(s);

    if (matched) {
      r.push({
        index: matched.index,
        length: matched[0].length
      });
    } else {
      break;
    }
  }

  return r;
};
/**
 * 使用标准规范的省略号
 * 要判断准确，还是必须一个一个处理
 * use-standard-ellipsis
 */


module.exports =
/*#__PURE__*/
function (_Plugin) {
  _inherits(_class, _Plugin);

  function _class() {
    _classCallCheck(this, _class);

    return _possibleConstructorReturn(this, _getPrototypeOf(_class).apply(this, arguments));
  }

  _createClass(_class, [{
    key: "pre",
    value: function pre() {}
  }, {
    key: "visitor",
    value: function visitor() {
      var _this = this;

      return {
        text: function text(ast) {
          var text = ast.node.value;
          var line = ast.node.position.start.line;
          var column = ast.node.position.start.column;
          findAllDotEllipsis(text).concat(findAllSingleEllipsis(text)).forEach(function (item) {
            _this.cfg.throwError({
              start: {
                line: line,
                column: column + item.index
              },
              end: {
                line: line,
                column: column + item.index + item.length
              },
              text: "'".concat(subErrorStr(text, item.index, showLength), "'"),
              ast: ast
            });
          });
        }
      };
    }
  }, {
    key: "post",
    value: function post() {}
  }], [{
    key: "type",
    get: function get() {
      return 'use-standard-ellipsis';
    }
  }]);

  return _class;
}(_astPlugin.Plugin);
